pkg/types/types.go

```diff
@@ -0,0 +1,104 @@
+	slsa "github.com/project-oak/transparent-release/pkg/intoto/slsa_provenance/v0.2"
+)
+
+// ValidatedProvenance wraps an intoto.Statement representing a valid SLSA provenance statement.
+// A provenance statement is valid if it contains a single subject, with a SHA256 hash.
+type ValidatedProvenance struct {
+	// The field is private so that invalid instances cannot be created.
+	provenance intoto.Statement
+}
+
+// NewValidatedProvenance validates the given provenance and returns an
+// instance of ValidatedProvenance wrapping it if it is valid, or an error
+// otherwise.
+func NewValidatedProvenance(provenance intoto.Statement) (*ValidatedProvenance, error) {
+	return &ValidatedProvenance{provenance: provenance}, nil
+}
+
+// GetBinarySHA256Digest returns the SHA256 digest of the subject.
+func (p *ValidatedProvenance) GetBinarySHA256Digest() string {
+	return p.provenance.Subject[0].Digest["sha256"]
+}
+
+// GetBinaryName returns the name of the subject.
+func (p *ValidatedProvenance) GetBinaryName() string {
+	return p.provenance.Subject[0].Name
+}
+
+// GetBuildType returns the build type of the provenance.
+func (p *ValidatedProvenance) GetBuildType() string {
+	return p.provenance.Predicate.(slsa.ProvenancePredicate).BuildType
+}
+
+// GetProvenance returns a partial copy of the provenance statement wrapped in this instance.
+// The partial copy guarantees that the validity condition will not be violated.
+func (p *ValidatedProvenance) GetProvenance() intoto.Statement {
+	subject := intoto.Subject{
+		Name:   p.provenance.Subject[0].Name,
+		Digest: intoto.DigestSet{"sha256": p.provenance.Subject[0].Digest["sha256"]},
+	}
+
+	statementHeader := intoto.StatementHeader{
+		Type:          p.provenance.Type,
+		PredicateType: p.provenance.PredicateType,
+		Subject:       []intoto.Subject{subject},
+	}
+
+	return intoto.Statement{
+		StatementHeader: statementHeader,
+		Predicate:       p.provenance.Predicate,
+	}
+}
+
+// ParseProvenanceData validates that the given bytes represent a valid SLSA provenance.
+// Returns an error if the bytes do not represent a valid JSON-encoded provenance statement.
+// Otherwise returns an instance of ValidatedProvenance.
+func ParseProvenanceData(statementBytes []byte) (*ValidatedProvenance, error) {
+	var statement intoto.Statement
+	if err := json.Unmarshal(statementBytes, &statement); err != nil {
+		return nil, fmt.Errorf("could not unmarshal the provenance file:\n%v", err)
+	}
+
+	if len(statement.Subject) != 1 || statement.Subject[0].Digest["sha256"] == "" {
+		return nil, fmt.Errorf("the provenance must have exactly one subject with a sha256 digest")
+	}
+
+	// statement.Predicate is now just a map, we have to parse it into an instance of slsa.ProvenancePredicate
+	predicateBytes, err := json.Marshal(statement.Predicate)
+	if err != nil {
+		return nil, fmt.Errorf("could not marshal Predicate map into JSON bytes: %v", err)
+	}
+
+	var predicate slsa.ProvenancePredicate
```

Comment:
I think, the downside is: at this point you are locked into slsav02.ProvenancePredicate. 

How will this change when there is a slsa v1.0?

Reply:
This should not change for slsa v1.0 I think. But we have to wait and see when the new format is out. 
Note that the files in `slsa_provenance/v02` are copied from the `slsa` repo, but we have full control on the structs. If needed, we can update these to a new common denominator.

Reply:
Here is the slsa v1 draf: https://github.com/slsa-framework/slsa/pull/525/files. Let's discuss this part separately.